android/engine/src/main/java/org/smartregister/fhircore/engine/domain/repository/Repository.kt

```diff
@@ -45,15 +47,41 @@ interface Repository {
   ): Long
 
   /**
-   * This function returns data used on the profile for the given [resourceId]. Profile
+   * This function returns baseResource used on the profile for the given [resourceId]. It returns
+   * this inside a container that is used to populate the remaining resources used in the profile.
+   * The remaining resources are loaded in [loadProfileRelatedAndSecondaryResources] Profile
    * configuration is identified by the [profileId] and contains the queries for filtering the
    * profile data. Data is loaded based on the [FhirResourceConfig]. When none is provided the
    * configurations identified by the [profileId] are used.
+   *
+   * @param profileId the ID of the profile to retrieve configurations from if fhirResourceConfig is
+   * null.
+   * @param resourceId the ID of the resource to get a baseResource for.
+   * @param fhirResourceConfig optional config.
+   * @param paramsList ActionParameters to customize functionality.
    */
-  suspend fun loadProfileData(
+  suspend fun loadProfileBaseResource(
     profileId: String,
     resourceId: String,
     fhirResourceConfig: FhirResourceConfig? = null,
     paramsList: Array<ActionParameter>?
   ): RepositoryResourceData?
+
+  /**
+   * This function loads the related resources and secondary resources into the [queryResult]. This
+   * function calls the [afterFetch] function after every resource is loaded. The [afterFetch]
+   * function should perform any processing that needs to be done on the newly loaded resources. The
+   * function also updates the [listResourceDataMapState] with the [ResourceData] for a specific
+   * view where the viewId is used as the map key
+   */
+  suspend fun loadProfileRelatedAndSecondaryResources(
+    queryResult: RepositoryResourceData.QueryResult.Search,
+    listResourceDataMapState: SnapshotStateMap<String, SnapshotStateList<ResourceData>>,
+    resourceData: ResourceData,
+    profileId: String,
+    resourceId: String,
+    fhirResourceConfig: FhirResourceConfig?,
+    paramsList: Array<ActionParameter>?,
+    afterFetch: suspend () -> Unit
```

Comment:
I also noticed this afterFetch function is called multiple times (as many as the number of items iterated in the LIST view). Internally it has a forEach function that searches for all the LIST properties of the profile view adding to the number of calls.

Reply:
Yes. It's a very dirty solution written with the assumption that n will not grow and the cost of repeating these operations is small enough compared to rewriting the calculations and resource loading to be accurate enough to recalculate only operations that affect the recently loaded resources